server/chat-plugins/sample-teams.ts

```diff
@@ -0,0 +1,515 @@
+						const team = category[teamName];
+						if (SampleTeams.checkPermissions(user, file.whitelist[query[0]] || [])) {
+							buf += `<button class="button" style="float:right" name="send" value="/sampleteams remove ${query[0]},${categoryName},${teamName}">Delete team</button>`;
+						}
+						buf += SampleTeams.formatTeam(teamName, team);
+					}
+					buf += `</details>`;
+					const index = Object.keys(file.teams[toID(query[0])]).indexOf(categoryName);
+					if (index !== Object.keys(file.teams[toID(query[0])]).length - 1) buf += `<hr />`;
+				}
+			} else if (SampleTeams.findCategory(query[0], query[1])) {
+				const categoryName = SampleTeams.findCategory(query[0], query[1])!;
+				buf += `<h3>Sample teams for ${SampleTeams.getFormatName(query[0])} in the ${categoryName} category</h3>`;
+				for (const teamName in file.teams[query[0]][categoryName]) {
+					const team = file.teams[query[0]][categoryName][teamName];
+					buf += SampleTeams.formatTeam(teamName, team);
+				}
+			}
+			buf += `</div>`;
+			return buf;
+		},
+		add(query, user, connection) {
+			this.title = `Sample Teams`;
+			const formats = Array.from(new Set(Dex.formats.all().map(format => (format.name.includes('(') ?
+				format.name.slice(0, format.name.indexOf('(')) : format.name).trim()))).map(formatName =>
+				Dex.formats.get(formatName)).filter(format =>
+				!format.name.includes('Custom') && format.effectType === 'Format' && !format.team &&
+				SampleTeams.checkPermissions(user, SampleTeams.whitelistedRooms(format.id) || []));
+			if (!formats.length) return `<div class="pad"><h2>Access denied.</h2></div>`;
+			let buf = `<div class="pad"><h2>Add a sample team</h2>`;
+			if (!query[0] || !Dex.formats.get(query[0]).exists) {
+				buf += `<h3>Pick a format</h3><ul>`;
+				for (const format of formats) {
+					buf += `<li>${formatFakeButton(`view-sampleteams-add-${format.id}`, format.name)}</button></li>`;
+				}
+				buf += `</ul>`;
+			}
+			if (!query[1] || !SampleTeams.findCategory(query[0], query[1]) || query[1] === 'addnewcategory') {
+				const name = SampleTeams.getFormatName(query[0]);
+				if (query[1] === 'addnewcategory') {
+					buf += `<h3>Add a category for ${name}</h3>`;
+					buf += `<form data-submitsend="/sampleteams addcategory ${query[0]},{categoryname}">`;
+					buf += `<input name="categoryname" />`;
+					buf += `<button class="button" type="submit">Add category</button></form>`;
+				} else {
+					buf += `<h3>Pick a category for ${name}</h3><ul>`;
+					for (const category of Object.keys(file.teams[query[0]] || {}) || []) {
+						buf += `<li style="padding-top:5px">${formatFakeButton(`view-sampleteams-add-${query[0]}-${toID(category)}-submit`, category)}</li>`;
+					}
+					buf += `<li style="padding-top:5px">${formatFakeButton(`view-sampleteams-add-${query[0]}-addnewcategory`, `<strong>Add new category</strong>`)}</li></ul>`;
+				}
+			}
+			if (query[2] === 'submit' && SampleTeams.findCategory(query[0], query[1])) {
```

Comment:
I doubt this will scale to the point where it matters, but you should probably use a variable for the result of `findCategory`